File: mahaam-api-java/src/main/java/mahaam/infra/monitor/LogRepo.java
package mahaam.infra.monitor;

import java.util.UUID;
import java.util.concurrent.CompletableFuture;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import mahaam.infra.Cache;
import mahaam.infra.DB;
import mahaam.infra.Log;
import mahaam.infra.Mapper;

public interface LogRepo {
	void create(UUID trafficId, String type, String message);
}

@ApplicationScoped
class DefaultLogRepo implements LogRepo {

	@Inject
	DB db;

	@Inject
	Cache cache;

	@Override
	public void create(UUID trafficId, String type, String message) {
		CompletableFuture.runAsync(
				() -> {
					try {
						String query = """
								INSERT INTO monitor.logs (traffic_id, type, message, node_ip, created_at)
								VALUES (:trafficId, :type, :message, :node_ip, current_timestamp)
								""";

						var params = Mapper.of(
								"trafficId", trafficId,
								"type", type,
								"message", message,
								"node_ip", cache.nodeIP());

						db.insert(query, params);
					} catch (Exception ex) {
						Log.error(
								"Unable to create log record:("
										+ type
										+ ": "
										+ message
										+ "). Cause: "
										+ ex.toString());
					}
				});
	}
}

File: mahaam-api-cs/Src/Infra/Monitor/LogRepo.cs

using System.Transactions;

namespace Mahaam.Infra.Monitoring;

public interface ILogRepo
{
	void Create(string type, string message, Guid? trafficId);
}

public class LogRepo(IDB db, ICache cache) : ILogRepo
{
	public void Create(string type, string message, Guid? trafficId)
	{
		var err = new
		{
			trafficId,
			type,
			message,
			node_ip = cache.NodeIP(),
		};
		var query = @"INSERT INTO monitor.logs (traffic_id, type, message, node_ip, created_at) 
			VALUES (@trafficId, @type, @message, @node_ip, current_timestamp)";

		Task.Run(() =>
			{
				try
				{
					using var scope = new TransactionScope(TransactionScopeOption.Suppress);
					db.Insert(query, err);
				}
				catch (Exception ex)
				{
					Serilog.Log.Error($"Unable to create log record:({type}: {message}). Cause: {ex.ToString()}");
				}
			}
		);

	}
}